servicetalk-encoding-netty/src/main/java/io/servicetalk/encoding/netty/NettyChannelContentCodec.java

```diff
@@ -0,0 +1,375 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.encoding.netty;
+
+import io.servicetalk.buffer.api.Buffer;
+import io.servicetalk.buffer.api.BufferAllocator;
+import io.servicetalk.buffer.api.CompositeBuffer;
+import io.servicetalk.buffer.netty.BufferUtils;
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Publisher;
+import io.servicetalk.encoding.api.CodecDecodingException;
+import io.servicetalk.encoding.api.CodecEncodingException;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.embedded.EmbeddedChannel;
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.MessageToByteEncoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Queue;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+
+import static io.netty.util.internal.PlatformDependent.throwException;
+import static io.servicetalk.buffer.api.ReadOnlyBufferAllocators.DEFAULT_RO_ALLOCATOR;
+import static io.servicetalk.buffer.netty.BufferUtils.getByteBufAllocator;
+import static io.servicetalk.buffer.netty.BufferUtils.newBufferFrom;
+import static io.servicetalk.concurrent.api.Single.succeeded;
+import static java.lang.String.valueOf;
+import static java.util.Objects.requireNonNull;
+
+final class NettyChannelContentCodec extends AbstractContentCodec {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NettyChannelContentCodec.class);
+    private static final Buffer END_OF_STREAM = DEFAULT_RO_ALLOCATOR.fromAscii(" ");
+    private static final int MAX_SIZE_FOR_MERGED_BUFFER = 1024; //1KiB
+
+    private final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier;
+    private final Supplier<ByteToMessageDecoder> decoderSupplier;
+
+    NettyChannelContentCodec(final CharSequence name,
+                             final Supplier<MessageToByteEncoder<ByteBuf>> encoderSupplier,
+                             final Supplier<ByteToMessageDecoder> decoderSupplier) {
+        super(name);
+        this.encoderSupplier = encoderSupplier;
+        this.decoderSupplier = decoderSupplier;
+    }
+
+    @Override
+    public Buffer encode(final Buffer src, final int offset, final int length, final BufferAllocator allocator) {
+        requireNonNull(src);
+        requireNonNull(allocator);
+
+        if (src.readableBytes() == 0) {
+            throw new CodecEncodingException(this, "No data to encode.");
+        }
+
+        if (offset < 0 || offset >= src.readableBytes()) {
+            throw new IllegalArgumentException("Invalid offset: " + offset +
+                    " (expected 0 - " + (src.readableBytes() - 1) + ")");
+        }
+
+        final int availableBytes = src.readableBytes() - offset;
+        if (length > availableBytes || length < 1) {
```

Comment:
You are taking the slice later, it will perform all the same checks for you. Why do we need to duplicate them here instead of taking the slice in the first line as proposed? Do the slice outside of the `try` block.

If I miss something and these checks are still necessary, consider moving them to a private method and share with `decode` method.

Reply:
Ok, agreed it makes less complicated.